File: src/agent/orchestrator/claude-code-subagent.ts
import {
  CLAUDE_CODE_MCP_SERVER_NAME,
  createMechaCoderMcpServer,
  getAllowedClaudeCodeTools,
} from "./claude-code-mcp.js";
import {
  AbortError,
  type PermissionMode,
  type McpServerConfig,
  type SDKAssistantMessageError,
  type HookCallback,
  type HookCallbackMatcher,
} from "@anthropic-ai/claude-agent-sdk";
import {
  type SubagentResult,
  type Subtask,
  type ClaudeCodePermissionMode,
  type OrchestratorEvent,
} from "./types.js";

type QueryFn = (input: unknown) => AsyncIterable<any>;

interface RetryConfig {
  maxRetries: number;
  initialDelayMs: number;
  maxDelayMs: number;
  backoffMultiplier: number;
}

export interface ClaudeCodeSubagentOptions {
  cwd: string;
  maxTurns?: number;
  systemPrompt?: string;
  permissionMode?: ClaudeCodePermissionMode | PermissionMode;
  queryFn?: QueryFn;
  openagentsDir?: string;
  allowedTools?: string[];
  mcpServers?: Record<string, unknown>;
  signal?: AbortSignal;
  timeoutMs?: number;
  onEvent?: (event: OrchestratorEvent) => void;
  /** Resume a prior Claude Code session if available */
  resumeSessionId?: string;
  /** Fork the resumed session to a new branch */
  forkSession?: boolean;
  /** Callback for streaming text output */
  onOutput?: (text: string) => void;
}

const isToolCallMessage = (message: any): message is { tool_calls?: Array<{ name?: string; input?: any }> } =>
  typeof message === "object" && message !== null && Array.isArray((message as any).tool_calls);

const isResultMessage = (message: any): message is { type?: string; subtype?: string } =>
  typeof message === "object" && message !== null && "type" in message;

const defaultBuildPrompt = (subtask: Subtask): string =>
  `## Subtask: ${subtask.id}

${subtask.description}

Focus on minimal, correct changes.`;

const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxRetries: 3,
  initialDelayMs: 1000,
  maxDelayMs: 32000,
  backoffMultiplier: 2,
};

const DEFAULT_TIMEOUT_MS = 50 * 60 * 1000; // 50 minutes

const assistantErrorMessages: Record<SDKAssistantMessageError, { message: string; suggestion?: string }> = {
  authentication_failed: {
    message: "Claude Code authentication failed",
    suggestion: "Set a valid ANTHROPIC_API_KEY before retrying",
  },
  billing_error: {
    message: "Claude Code billing error",
    suggestion: "Check billing status for Anthropic API access",
  },
  rate_limit: {
    message: "Claude Code hit rate limits",
    suggestion: "Back off and retry later or fall back to the minimal subagent",
  },
  invalid_request: {
    message: "Claude Code rejected the request",
    suggestion: "Inspect Claude Code inputs and configuration for invalid parameters",
  },
  server_error: {
    message: "Claude Code server error",
    suggestion: "Retry after a short delay or fall back to the minimal subagent",
  },
  unknown: { message: "Claude Code returned an unknown error" },
};

const describeAssistantError = (errorType?: SDKAssistantMessageError): { message: string; suggestion?: string } => {
  if (!errorType) return { message: "Claude Code reported an error" };
  return assistantErrorMessages[errorType] ?? { message: `Claude Code reported ${errorType}` };
};

const createAbortController = (signal?: AbortSignal, timeoutMs?: number) => {
  const controller = new AbortController();
  let abortReason: string | undefined;
  let timeoutId: ReturnType<typeof setTimeout> | undefined;

  const propagateAbort = () => {
    if (controller.signal.aborted) return;
    abortReason =
      typeof signal?.reason === "string"
        ? signal.reason
        : signal?.reason instanceof Error
          ? signal.reason.message
          : "Claude Code run aborted";
    controller.abort(new Error(abortReason));
  };

  if (signal?.aborted) {
    propagateAbort();
  } else if (signal) {
    signal.addEventListener("abort", propagateAbort);
  }

  const timeout = timeoutMs ?? DEFAULT_TIMEOUT_MS;
  if (timeout && timeout > 0) {
    timeoutId = setTimeout(() => {
      abortReason = `Claude Code timed out after ${timeout}ms`;
      controller.abort(new Error(abortReason));
    }, timeout);
  }

  const cleanup = () => {
    if (timeoutId) clearTimeout(timeoutId);
    if (signal) signal.removeEventListener("abort", propagateAbort);
  };

  return {
    controller,
    getAbortReason: () => abortReason,
    cleanup,
  };
};

/**
 * Check if an error is retryable (rate limit or server error)
 */
const isRetryableError = (errorType?: SDKAssistantMessageError | string): boolean => {
  return errorType === "rate_limit" || errorType === "server_error";
};

/**
 * Check if an error is a fatal auth error (should stop immediately)
 */
const isFatalAuthError = (errorType?: SDKAssistantMessageError | string): boolean => {
  return errorType === "authentication_failed" || errorType === "billing_error";
};

/**
 * Calculate delay for exponential backoff
 */
const calculateBackoffDelay = (
  attempt: number,
  config: RetryConfig = DEFAULT_RETRY_CONFIG
): number => {
  const delay = config.initialDelayMs * Math.pow(config.backoffMultiplier, attempt);
  return Math.min(delay, config.maxDelayMs);
};

/**
 * Sleep for a specified number of milliseconds
 */
const sleep = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Create SDK hooks for observability and event emission
 *
 * NOTE: Hooks are for side effects (logging, events) only.
 * Tool/file tracking is done via message parsing for consistency across real SDK and test mocks.
 */
const createClaudeCodeHooks = (
  onEvent?: (event: OrchestratorEvent) => void
): Partial<Record<string, HookCallbackMatcher[]>> => {
  const postToolUseHook: HookCallback = async (input) => {
    if (input.hook_event_name === "PostToolUse") {
      // Emit event if callback provided
      if (onEvent) {
        onEvent({
          type: "subtask_tool_use",
          toolName: input.tool_name,
          input: input.tool_input,
        } as any); // Cast to any since this event type doesn't exist yet
      }
    }
    return { continue: true };
  };

  const sessionEndHook: HookCallback = async (input) => {
    if (input.hook_event_name === "SessionEnd") {
      // Emit event if callback provided
      if (onEvent) {
        onEvent({
          type: "claude_code_session_end",
          reason: input.reason,
          sessionId: input.session_id,
        } as any); // Cast to any since this event type doesn't exist yet
      }
    }
    return { continue: true };
  };

  return {
    PostToolUse: [{ hooks: [postToolUseHook] }],
    SessionEnd: [{ hooks: [sessionEndHook] }],
  };
};

/**
 * Run a subtask using Claude Code Agent SDK's query() streaming API with retry logic.
 * Implements:
 * - Exponential backoff for rate limits
 * - Retry for server errors
 * - Immediate failure for auth errors
 * - Logs all errors to blockers for progress.md
 */
export const runClaudeCodeSubagent = async (
  subtask: Subtask,
  options: ClaudeCodeSubagentOptions
): Promise<SubagentResult> => {
  const query = options.queryFn ?? (await import("@anthropic-ai/claude-agent-sdk")).query;

  const filesModified: Set<string> = new Set();
  const toolsUsed: Map<string, number> = new Map();
  const blockers: string[] = [];
  const assistantMessages: string[] = [];
  const suggestedNextSteps: string[] = [];
  let success = false;
  let error: string | undefined;
  let turns = 0;
  let lastErrorType: SDKAssistantMessageError | string | undefined;
  let totalCostUsd: number | undefined;
  let usage: any | undefined;
  let sessionId: string | undefined = options.resumeSessionId;

  const mcpServers: Record<string, McpServerConfig> =
    (options.mcpServers as Record<string, McpServerConfig> | undefined) ??
    {
      [CLAUDE_CODE_MCP_SERVER_NAME]: createMechaCoderMcpServer({
        openagentsDir: options.openagentsDir ?? "",
      }) as McpServerConfig,
    };

  const allowedTools = options.allowedTools ?? getAllowedClaudeCodeTools();

  const retryConfig = DEFAULT_RETRY_CONFIG;
  let retryAttempt = 0;
  const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;

  // Create hooks for observability and event emission
  const hooks = createClaudeCodeHooks(options.onEvent);

  // Retry loop for rate limits and server errors
  while (retryAttempt <= retryConfig.maxRetries) {
    const { controller, getAbortReason, cleanup } = createAbortController(options.signal, timeoutMs);

    try {
      for await (const message of query({
        prompt: defaultBuildPrompt(subtask),
        options: {
          cwd: options.cwd,
          // Use Claude Code's system prompt with CLAUDE.md context
          systemPrompt: options.systemPrompt ?? { type: "preset", preset: "claude_code" },
          maxTurns: options.maxTurns ?? 300,
          ...(options.permissionMode ? { permissionMode: options.permissionMode } : {}),
          ...(options.resumeSessionId ? { resume: options.resumeSessionId } : {}),
          ...(options.forkSession ? { forkSession: true } : {}),
          mcpServers,
          allowedTools,
          abortController: controller,
          hooks,
          settingSources: ["project"], // Load CLAUDE.md for project context
          includePartialMessages: !!options.onOutput, // Enable streaming if callback provided
        },
      })) {
        // Handle streaming partial messages
        if ((message as any).type === "stream_event" && options.onOutput) {
          const event = (message as any).event;
          // Extract text delta from content_block_delta events
          if (event?.type === "content_block_delta" && event?.delta?.type === "text_delta") {
            options.onOutput(event.delta.text);
          }
          // Add newline at end of text blocks
          if (event?.type === "content_block_stop") {
            options.onOutput("\n");
          }
        }

        // Log all non-stream messages as JSON for visibility
        if (options.onOutput && (message as any).type !== "stream_event") {
          const msgType = (message as any).type || (message as any).role || "unknown";
          // Log tool use from assistant messages
          if ((message as any).role === "assistant" && Array.isArray((message as any).content)) {
            for (const block of (message as any).content) {
              if (block.type === "tool_use") {
                const toolJson = JSON.stringify({ tool: block.name, id: block.id, input: block.input }, null, 0);
                options.onOutput(`\n[TOOL_USE] ${toolJson}\n`);
              }
            }
          }
          // Log tool results
          if ((message as any).role === "user" && Array.isArray((message as any).content)) {
            for (const block of (message as any).content) {
              if (block.type === "tool_result") {
                const truncatedContent = typeof block.content === "string" && block.content.length > 200
                  ? block.content.slice(0, 200) + "..."
                  : block.content;
                const resultJson = JSON.stringify({ tool_result: block.tool_use_id, content: truncatedContent }, null, 0);
                options.onOutput(`[TOOL_RESULT] ${resultJson}\n`);
              }
            }
          }
          // Log result/system messages
          if (msgType === "result" || msgType === "system") {
            const resultJson = JSON.stringify(message, null, 0);
            options.onOutput(`[${msgType.toUpperCase()}] ${resultJson}\n`);
          }
        }

        // Legacy: Log tool calls as JSON (for messages with tool_calls array)
        if (isToolCallMessage(message) && message.tool_calls && options.onOutput) {
          for (const call of message.tool_calls) {
            if (call?.name) {
              const toolJson = JSON.stringify({ tool: call.name, input: call.input });
              options.onOutput(`\n[TOOL] ${toolJson}\n`);
            }
          }
        }
        // Track tool calls from messages
        // NOTE: Hooks provide event emission, but message parsing tracks usage/files
        // for consistency across real SDK and test mocks
        if (isToolCallMessage(message) && message.tool_calls) {
          for (const call of message.tool_calls) {
            const name = call?.name;
            if (name) {
              toolsUsed.set(name, (toolsUsed.get(name) || 0) + 1);
            }

            const filePath = call?.input?.file_path || call?.input?.path;
            if (filePath && (name === "Edit" || name === "Write")) {
              filesModified.add(String(filePath));
            }
          }
        }

        // Track tool_use blocks from assistant content (SDK message format)
        if ((message as any).role === "assistant" && Array.isArray((message as any).content)) {
          for (const block of (message as any).content) {
            if (block.type === "tool_use" && block.name) {
              toolsUsed.set(block.name, (toolsUsed.get(block.name) || 0) + 1);
              const filePath = block.input?.file_path || block.input?.path;
              if (filePath && (block.name === "Edit" || block.name === "Write")) {
                filesModified.add(String(filePath));
              }
            }
          }
        }

        // Capture assistant messages for context
        if ((message as any).role === "assistant" && (message as any).content) {
          const content = String((message as any).content);
          if (content.length > 0 && content.length < 500) {
            assistantMessages.push(content);
          }
        }

        // Capture session IDs for resumption
        const messageSessionId = (message as any)?.session_id;
        if (typeof messageSessionId === "string") {
          sessionId = messageSessionId;
        }

        // Capture SDK error messages with error types
        if ((message as any).error) {
          lastErrorType = (message as any).error as SDKAssistantMessageError;
          const assistantError = describeAssistantError(lastErrorType as SDKAssistantMessageError);
          error = assistantError.message;
          blockers.push(assistantError.message);
          if (assistantError.suggestion) {
            suggestedNextSteps.push(assistantError.suggestion);
          }

          // Check for fatal auth errors - stop immediately
          if (isFatalAuthError(lastErrorType)) {
            success = false;
            const fatalMessage =
              assistantError.suggestion && !assistantError.message.includes(assistantError.suggestion)
                ? `${assistantError.message}. ${assistantError.suggestion}`
                : assistantError.message;
            error = fatalMessage;
            blockers.push(fatalMessage);
            controller.abort(new Error(fatalMessage));
            break; // Exit message loop
          }
        } else if ((message as any).isError) {
          const errorMsg = String((message as any).message || "Unknown error");
          blockers.push(errorMsg);
        }

        if (isResultMessage(message) && message.type === "result") {
          const permissionDenials = (message as any).permission_denials;
          if (Array.isArray(permissionDenials)) {
            for (const denial of permissionDenials) {
              const toolName = denial?.tool_name;
              const reason = toolName ? `Permission denied for ${toolName}` : "Permission denied for tool use";
              blockers.push(reason);
              suggestedNextSteps.push("Update Claude Code permissions or enable bypassPermissions for automation runs");
            }
          }

          turns = (message as any).turns ?? (message as any).num_turns ?? turns;

          // Capture usage and cost data
          if ((message as any).usage) {
            usage = (message as any).usage;
          }
          if (typeof (message as any).total_cost_usd === "number") {
            totalCostUsd = (message as any).total_cost_usd;
          }

          if (message.subtype === "success") {
            success = true;
          } else {
            success = false;
            error = `Claude Code finished with: ${message.subtype ?? "unknown"}`;
            if (error) blockers.push(error);
          }
          break;
        }
      }

      // If successful or fatal error, break retry loop
      if (controller.signal.aborted) {
        success = false;
        if (!error) {
          error = getAbortReason() ?? "Claude Code run aborted";
        }
        blockers.push(error);
        break;
      }

      if (success || isFatalAuthError(lastErrorType)) {
        break;
      }

      // If we have a retryable error and haven't exhausted retries, retry
      if (lastErrorType && isRetryableError(lastErrorType) && retryAttempt < retryConfig.maxRetries) {
        const delay = calculateBackoffDelay(retryAttempt, retryConfig);
        const retryMsg = `Retrying after ${lastErrorType} (attempt ${retryAttempt + 1}/${retryConfig.maxRetries}, delay ${delay}ms)`;
        blockers.push(retryMsg);
        await sleep(delay);
        retryAttempt++;
        // Reset error state for retry
        lastErrorType = undefined;
        continue;
      }

      // No retryable error or retries exhausted - exit loop
      break;
    } catch (e: any) {
      success = false;
      const abortReason = getAbortReason();
      if (e instanceof AbortError || e?.name === "AbortError") {
        error = abortReason ?? e?.message ?? "Claude Code run aborted";
      } else if (typeof e?.status === "number" && (e.status === 401 || e.status === 403)) {
        error = "Claude Code authentication failed (401/403)";
        suggestedNextSteps.push("Verify ANTHROPIC_API_KEY and SDK authentication");
      } else if (typeof e?.status === "number" && e.status === 429) {
        error = "Claude Code rate limited (429)";
        suggestedNextSteps.push("Back off and retry after rate limit resets");
        lastErrorType = "rate_limit";
      } else {
        error = e?.message || String(e);
      }
      const catchMsg = `Exception during query: ${error}`;
      blockers.push(catchMsg);

      if (abortReason) {
        break;
      }

      // Check if exception might be retryable (network errors, etc)
      const errorLower = (error ?? "").toLowerCase();
      const isNetworkError =
        errorLower.includes("network") ||
        errorLower.includes("timeout") ||
        errorLower.includes("econnrefused");

      if (isNetworkError && retryAttempt < retryConfig.maxRetries) {
        const delay = calculateBackoffDelay(retryAttempt, retryConfig);
        const retryMsg = `Retrying after network error (attempt ${retryAttempt + 1}/${retryConfig.maxRetries}, delay ${delay}ms)`;
        blockers.push(retryMsg);
        await sleep(delay);
        retryAttempt++;
        continue;
      }

      // Non-retryable exception or retries exhausted
      break;
    } finally {
      cleanup();
    }
  }

  if (success) {
    error = undefined;
  }

  if (!success && !error) {
    error = "Claude Code ended without a result";
    blockers.push(error);
  }

  // Add hint if retries were exhausted (signals orchestrator to fallback)
  if (!success && retryAttempt >= retryConfig.maxRetries && lastErrorType) {
    const exhaustedMsg = `Retries exhausted after ${retryAttempt} attempts. Consider fallback to minimal subagent.`;
    blockers.push(exhaustedMsg);
    if (!error) {
      error = exhaustedMsg;
    }
  }

  const result: SubagentResult = {
    success,
    subtaskId: subtask.id,
    filesModified: Array.from(filesModified),
    turns,
    agent: "claude-code",
  };

  if (sessionId) {
    result.claudeCodeSessionId = sessionId;
    result.sessionMetadata = {
      ...(result.sessionMetadata ?? {}),
      sessionId,
      ...(options.resumeSessionId && options.forkSession && sessionId !== options.resumeSessionId
        ? { forkedFromSessionId: options.resumeSessionId }
        : {}),
    };

    if (options.resumeSessionId && options.forkSession && sessionId !== options.resumeSessionId) {
      result.claudeCodeForkedFromSessionId = options.resumeSessionId;
    }

    // Log session for debugging
    console.log(`[Claude Code] Session: ${sessionId}${options.resumeSessionId ? ` (resumed from ${options.resumeSessionId})` : " (new)"}`);
  }

  if (error) {
    result.error = error;
  }

  // Add session metadata for progress.md bridging
  if (
    toolsUsed.size > 0 ||
    blockers.length > 0 ||
    assistantMessages.length > 0 ||
    suggestedNextSteps.length > 0 ||
    usage ||
    totalCostUsd !== undefined
  ) {
    const sessionMetadata = { ...(result.sessionMetadata ?? {}) };

    if (toolsUsed.size > 0) {
      sessionMetadata.toolsUsed = Object.fromEntries(toolsUsed);
    }

    if (blockers.length > 0) {
      sessionMetadata.blockers = blockers;
    }

    if (suggestedNextSteps.length > 0) {
      sessionMetadata.suggestedNextSteps = Array.from(new Set(suggestedNextSteps));
    }

    // Use the last assistant message as summary
    if (assistantMessages.length > 0) {
      sessionMetadata.summary = assistantMessages[assistantMessages.length - 1];
    }

    // Add token usage and cost data
    if (usage) {
      sessionMetadata.usage = {
        inputTokens: usage.input_tokens ?? usage.inputTokens,
        outputTokens: usage.output_tokens ?? usage.outputTokens,
        cacheReadInputTokens: usage.cache_read_input_tokens ?? usage.cacheReadInputTokens,
        cacheCreationInputTokens: usage.cache_creation_input_tokens ?? usage.cacheCreationInputTokens,
      };
    }

    if (totalCostUsd !== undefined) {
      sessionMetadata.totalCostUsd = totalCostUsd;
    }

    result.sessionMetadata = sessionMetadata;
  }

  return result;
};

File: crates/agent_openagents_backup/src/claude_code_subagent.rs
//! Claude Code Subagent
//!
//! Integration with Claude Code Agent SDK for running subtasks.
//! Implements:
//! - Exponential backoff for rate limits
//! - Retry for server errors
//! - Immediate failure for auth errors
//! - Session management and resumption

use crate::error::{AgentError, AgentResult};
use crate::types::{AgentType, SubagentResult, Subtask, SubtaskStatus};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// ─────────────────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────────────────

/// Retry configuration for Claude Code
#[derive(Debug, Clone)]
pub struct RetryConfig {
    /// Maximum number of retries
    pub max_retries: u32,
    /// Initial delay in milliseconds
    pub initial_delay_ms: u64,
    /// Maximum delay in milliseconds
    pub max_delay_ms: u64,
    /// Backoff multiplier
    pub backoff_multiplier: f64,
}

impl Default for RetryConfig {
    fn default() -> Self {
        Self {
            max_retries: 3,
            initial_delay_ms: 1000,
            max_delay_ms: 32000,
            backoff_multiplier: 2.0,
        }
    }
}

/// Claude Code subagent options
#[derive(Debug, Clone)]
pub struct ClaudeCodeSubagentOptions {
    /// Working directory
    pub cwd: String,
    /// Maximum turns
    pub max_turns: Option<u32>,
    /// System prompt override
    pub system_prompt: Option<String>,
    /// Permission mode
    pub permission_mode: Option<String>,
    /// OpenAgents directory
    pub openagents_dir: Option<String>,
    /// Allowed tools
    pub allowed_tools: Option<Vec<String>>,
    /// Timeout in milliseconds
    pub timeout_ms: Option<u64>,
    /// Resume a prior Claude Code session
    pub resume_session_id: Option<String>,
    /// Fork the resumed session
    pub fork_session: bool,
    /// Additional context (e.g., AGENTS.md)
    pub additional_context: Option<String>,
    /// Reflections from previous failures
    pub reflections: Option<String>,
    /// TB run ID for ATIF
    pub run_id: Option<String>,
}

impl Default for ClaudeCodeSubagentOptions {
    fn default() -> Self {
        Self {
            cwd: ".".to_string(),
            max_turns: Some(300),
            system_prompt: None,
            permission_mode: None,
            openagents_dir: None,
            allowed_tools: None,
            timeout_ms: Some(50 * 60 * 1000), // 50 minutes
            resume_session_id: None,
            fork_session: false,
            additional_context: None,
            reflections: None,
            run_id: None,
        }
    }
}

/// SDK assistant message error types
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SDKAssistantMessageError {
    /// Authentication failed
    AuthenticationFailed,
    /// Billing error
    BillingError,
    /// Rate limit hit
    RateLimit,
    /// Invalid request
    InvalidRequest,
    /// Server error
    ServerError,
    /// Unknown error
    Unknown,
}

/// Error message descriptions
#[derive(Debug, Clone)]
pub struct ErrorDescription {
    /// Error message
    pub message: String,
    /// Suggestion for resolution
    pub suggestion: Option<String>,
}


// ─────────────────────────────────────────────────────────────────────────────
// Helper Functions
// ─────────────────────────────────────────────────────────────────────────────

/// Get error message descriptions
pub fn describe_assistant_error(error_type: Option<SDKAssistantMessageError>) -> ErrorDescription {
    match error_type {
        Some(SDKAssistantMessageError::AuthenticationFailed) => ErrorDescription {
            message: "Claude Code authentication failed".to_string(),
            suggestion: Some("Set a valid ANTHROPIC_API_KEY before retrying".to_string()),
        },
        Some(SDKAssistantMessageError::BillingError) => ErrorDescription {
            message: "Claude Code billing error".to_string(),
            suggestion: Some("Check billing status for Anthropic API access".to_string()),
        },
        Some(SDKAssistantMessageError::RateLimit) => ErrorDescription {
            message: "Claude Code hit rate limits".to_string(),
            suggestion: Some(
                "Back off and retry later or fall back to the minimal subagent".to_string(),
            ),
        },
        Some(SDKAssistantMessageError::InvalidRequest) => ErrorDescription {
            message: "Claude Code rejected the request".to_string(),
            suggestion: Some(
                "Inspect Claude Code inputs and configuration for invalid parameters".to_string(),
            ),
        },
        Some(SDKAssistantMessageError::ServerError) => ErrorDescription {
            message: "Claude Code server error".to_string(),
            suggestion: Some(
                "Retry after a short delay or fall back to the minimal subagent".to_string(),
            ),
        },
        Some(SDKAssistantMessageError::Unknown) | None => ErrorDescription {
            message: "Claude Code reported an error".to_string(),
            suggestion: None,
        },
    }
}

/// Check if an error is retryable
pub fn is_retryable_error(error_type: Option<SDKAssistantMessageError>) -> bool {
    matches!(
        error_type,
        Some(SDKAssistantMessageError::RateLimit) | Some(SDKAssistantMessageError::ServerError)
    )
}

/// Check if an error is a fatal auth error
pub fn is_fatal_auth_error(error_type: Option<SDKAssistantMessageError>) -> bool {
    matches!(
        error_type,
        Some(SDKAssistantMessageError::AuthenticationFailed)
            | Some(SDKAssistantMessageError::BillingError)
    )
}

/// Calculate delay for exponential backoff
pub fn calculate_backoff_delay(attempt: u32, config: &RetryConfig) -> u64 {
    let delay =
        (config.initial_delay_ms as f64) * config.backoff_multiplier.powi(attempt as i32);
    (delay as u64).min(config.max_delay_ms)
}

/// Build prompt for subtask
pub fn build_prompt(
    subtask: &Subtask,
    additional_context: Option<&str>,
    reflections: Option<&str>,
    cwd: Option<&str>,
) -> String {
    let mut prompt = String::new();

    // Prepend additional context if provided
    if let Some(ctx) = additional_context {
        prompt.push_str("## Project Context\n\n");
        prompt.push_str(ctx);
        prompt.push_str("\n\n---\n\n");
    }

    // Worktree guidance if running in isolated worktree
    if let Some(cwd_path) = cwd {
        if cwd_path.contains("/.worktrees/") {
            prompt.push_str("## IMPORTANT: Worktree Isolation\n\n");
            prompt.push_str(&format!(
                "You are running in an ISOLATED git worktree at:\n`{}`\n\n",
                cwd_path
            ));
            prompt.push_str("This is your workspace. ALL your work happens here:\n");
            prompt.push_str("- Make all file changes in this worktree directory\n");
            prompt.push_str("- Run all commands (tests, typecheck, git) in this worktree\n");
            prompt.push_str("- DO NOT try to work in \"the main repo\" or switch directories\n");
            prompt.push_str("- The worktree IS the repository - it has all project files\n");
            prompt.push_str("- Your changes will be merged to main after you complete\n\n");
            prompt.push_str(
                "Stay in this worktree for all operations. This isolation prevents conflicts with other agents.\n\n",
            );
            prompt.push_str("---\n\n");
        }
    }

    // Subtask description
    prompt.push_str(&format!("## Subtask: {}\n\n", subtask.id));
    prompt.push_str(&subtask.description);
    prompt.push_str("\n\nFocus on minimal, correct changes.");

    // Include failure context if this is a retry
    if let (Some(failure_count), Some(last_failure)) =
        (subtask.failure_count, &subtask.last_failure_reason)
    {
        if failure_count > 0 {
            prompt.push_str(&format!(
                "\n\n## IMPORTANT: Previous Attempt Failed\n\n\
                This subtask has failed {} time(s). The last failure was:\n```\n{}\n```\n\n\
                You MUST address this error before proceeding. Do NOT repeat the same approach that caused the failure.\n\
                Run `bun run typecheck` and `bun test` to verify your changes pass before completing.",
                failure_count, last_failure
            ));
        }
    }

    // Include reflections from previous failures
    if let Some(ref_text) = reflections {
        if !ref_text.trim().is_empty() {
            prompt.push_str(&format!(
                "\n\n## Learning from Previous Failures\n\n\
                The following reflections were generated from previous failed attempts. \
                Use these insights to avoid repeating the same mistakes:\n\n{}",
                ref_text
            ));
        }
    }

    prompt
}

// ─────────────────────────────────────────────────────────────────────────────
// Main Entry Point
// ─────────────────────────────────────────────────────────────────────────────

/// Run a subtask using Claude Code Agent SDK
///
/// NOTE: This is a stub implementation. Actual SDK integration requires
/// either FFI bindings or a native Rust SDK.
pub fn run_claude_code_subagent(
    subtask: &Subtask,
    options: &ClaudeCodeSubagentOptions,
) -> AgentResult<SubagentResult> {
    // Build the prompt
    let _prompt = build_prompt(
        subtask,
        options.additional_context.as_deref(),
        options.reflections.as_deref(),
        Some(&options.cwd),
    );

    // TODO: Implement actual SDK integration
    // For now, return a placeholder result indicating SDK not available
    Ok(SubagentResult {
        success: false,
        subtask_id: subtask.id.clone(),
        files_modified: Vec::new(),
        turns: 0,
        error: Some("Claude Code SDK integration not yet implemented in Rust".to_string()),
        agent: Some(AgentType::ClaudeCode),
        claude_code_session_id: None,
        claude_code_forked_from_session_id: None,
        token_usage: None,
        verification_outputs: None,
        session_metadata: None,
        learning_metrics: None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::SessionMetadata;

    #[test]
    fn test_retry_config_default() {
        let config = RetryConfig::default();
        assert_eq!(config.max_retries, 3);
        assert_eq!(config.initial_delay_ms, 1000);
        assert_eq!(config.backoff_multiplier, 2.0);
    }

    #[test]
    fn test_calculate_backoff_delay() {
        let config = RetryConfig::default();

        // First retry: 1000 * 2^0 = 1000
        assert_eq!(calculate_backoff_delay(0, &config), 1000);

        // Second retry: 1000 * 2^1 = 2000
        assert_eq!(calculate_backoff_delay(1, &config), 2000);

        // Third retry: 1000 * 2^2 = 4000
        assert_eq!(calculate_backoff_delay(2, &config), 4000);
    }

    #[test]
    fn test_calculate_backoff_delay_max() {
        let config = RetryConfig {
            initial_delay_ms: 1000,
            max_delay_ms: 5000,
            backoff_multiplier: 2.0,
            ..Default::default()
        };

        // 1000 * 2^3 = 8000, but capped at 5000
        assert_eq!(calculate_backoff_delay(3, &config), 5000);
    }

    #[test]
    fn test_is_retryable_error() {
        assert!(is_retryable_error(Some(SDKAssistantMessageError::RateLimit)));
        assert!(is_retryable_error(Some(SDKAssistantMessageError::ServerError)));
        assert!(!is_retryable_error(Some(
            SDKAssistantMessageError::AuthenticationFailed
        )));
        assert!(!is_retryable_error(None));
    }

    #[test]
    fn test_is_fatal_auth_error() {
        assert!(is_fatal_auth_error(Some(
            SDKAssistantMessageError::AuthenticationFailed
        )));
        assert!(is_fatal_auth_error(Some(SDKAssistantMessageError::BillingError)));
        assert!(!is_fatal_auth_error(Some(SDKAssistantMessageError::RateLimit)));
        assert!(!is_fatal_auth_error(None));
    }

    #[test]
    fn test_describe_assistant_error() {
        let desc = describe_assistant_error(Some(SDKAssistantMessageError::RateLimit));
        assert!(desc.message.contains("rate limit"));
        assert!(desc.suggestion.is_some());
    }

    #[test]
    fn test_describe_assistant_error_unknown() {
        let desc = describe_assistant_error(None);
        assert!(desc.message.contains("error"));
        assert!(desc.suggestion.is_none());
    }

    #[test]
    fn test_build_prompt_basic() {
        let subtask = Subtask {
            id: "sub-1".to_string(),
            description: "Fix the bug".to_string(),
            status: SubtaskStatus::Pending,
            ..Default::default()
        };

        let prompt = build_prompt(&subtask, None, None, None);
        assert!(prompt.contains("sub-1"));
        assert!(prompt.contains("Fix the bug"));
    }

    #[test]
    fn test_build_prompt_with_context() {
        let subtask = Subtask {
            id: "sub-1".to_string(),
            description: "Fix the bug".to_string(),
            status: SubtaskStatus::Pending,
            ..Default::default()
        };

        let prompt = build_prompt(&subtask, Some("Project uses Bun"), None, None);
        assert!(prompt.contains("Project Context"));
        assert!(prompt.contains("Project uses Bun"));
    }

    #[test]
    fn test_build_prompt_with_worktree() {
        let subtask = Subtask {
            id: "sub-1".to_string(),
            description: "Fix the bug".to_string(),
            status: SubtaskStatus::Pending,
            ..Default::default()
        };

        let prompt = build_prompt(&subtask, None, None, Some("/repo/.worktrees/task-1"));
        assert!(prompt.contains("Worktree Isolation"));
        assert!(prompt.contains("/.worktrees/"));
    }

    #[test]
    fn test_build_prompt_with_failure() {
        let subtask = Subtask {
            id: "sub-1".to_string(),
            description: "Fix the bug".to_string(),
            status: SubtaskStatus::Pending,
            failure_count: Some(2),
            last_failure_reason: Some("Type error in line 42".to_string()),
            ..Default::default()
        };

        let prompt = build_prompt(&subtask, None, None, None);
        assert!(prompt.contains("Previous Attempt Failed"));
        assert!(prompt.contains("Type error in line 42"));
    }

    #[test]
    fn test_build_prompt_with_reflections() {
        let subtask = Subtask {
            id: "sub-1".to_string(),
            description: "Fix the bug".to_string(),
            status: SubtaskStatus::Pending,
            ..Default::default()
        };

        let prompt = build_prompt(&subtask, None, Some("Always run tests before committing"), None);
        assert!(prompt.contains("Learning from Previous Failures"));
        assert!(prompt.contains("Always run tests"));
    }

    #[test]
    fn test_session_metadata_serialization() {
        let metadata = SessionMetadata {
            session_id: Some("sess-123".to_string()),
            tools_used: Some([("Edit".to_string(), 5)].into_iter().collect()),
            ..Default::default()
        };

        let json = serde_json::to_string(&metadata).unwrap();
        assert!(json.contains("sess-123"));
        assert!(json.contains("Edit"));
    }

    #[test]
    fn test_claude_code_subagent_options_default() {
        let options = ClaudeCodeSubagentOptions::default();
        assert_eq!(options.max_turns, Some(300));
        assert!(!options.fork_session);
    }
}
